### DISCUSSION TITLE
Expand lambdas and method body definitions to allow any block like statements

### DISCUSSION THREAD

**Discussion by br3nt** [dotnet/csharplang#6885] [Language Ideas]
This proposal has come about because @CyrusNajmabadi has pointed out that the proposal from #6870 can be generalised to include all blocks.

The proposal is that lambdas and method body definitions be expanded to include block like statements such as selection statements (`if`, `switch`), iteration statements (`while`, `do`, `for`, `foreach`), `try`/`catch`/`finally` statements, `checked` and `unchecked` statements, `lock` statements, and `using` statements, and anything else that may fit in this category.

The general pattern being considered is:

```c#
public void Example() =>
    <block-statement>
```

and

```c#
public void Example()
{
    var list = new List<int>();

    var mapped = list.Select(i =>
        <block-statement>
    );
}
```

and

```c#
Func<int> example = () =>
    <block-statement>
```

Lets consider some examples using some of the block statement keywords just to get an idea...

#### `if/else`

Or,
```c#
public void MethodIf() =>
    if (condition)
    {
        // code
    }
    else
    {
        // ...
    }
```

Or, 

```c#
public void MethodIf() => if (condition) // code
```

Or,

```c#
collection.Select(innerArray =>
    if (condition)
    {
        // code
    }
    else
    {
        // ...
    });
```

Or,

```c#
Func<int> example = () =>
    if (condition)
    {
        return 1;
    }
    else
    {
        return 2;
    });
```

#### `for`

```c#
public IEnumerable<string> MethodForEach() =>
    for (var i = 0; i < 10; i++)
    {
        yield return i * 2;
    }
```

Or,
```c#
public IEnumerable<string> MethodForEach() => for (var i = 0; i < 10; i++) yield return i * 2;
```

Or,
```c#
Func<int> example = (count) =>
    for (var i = 0; i < count; i++)
    {
        yield return i * 2;
    }
```

Or,
```c#
Func<int> example = (count) => for (var i = 0; i < count; i++) yield return i * 2;
```

Hopefully, that's enough examples to get the idea.

It might be worthwhile to also consider implicit returns, like what is being done with the switch expression, but I'm not sure if that would work in all cases. I think the return and yield return works just fine.

**Comment by CyrusNajmabadi** [comment]
I'll champion this. Will attempt to get a read from teh rest of the LDG the next time we're reviewing things. Not sure how it will be viewed, but there may be enough interest here :)

**Comment by En3Tho** [comment]
Is the sole point of this just saving 2 brackets solely in case of single statement?

**Reply by CyrusNajmabadi** [reply]
@En3Tho Yup.  But the same was true when we added expression-bodied members, and supported expression-statements.  This just expands that same support out to all statements.

**Reply by foxesknow** [reply]
> @En3Tho Yup. But the same was true when we added expression-bodied members, and supported expression-statements. This just expands that same support out to all statements.

Adding expression-bodied members was a mistake (in my opinion). We've ended up with an additional way of writing methods, which if you need to extend the method often means you need to reformat from an expression-bodied member to a regular memory. 

Since it's introduction I've seen developers write convoluted code in order to needlessly make a method a single expression.

If the plan is to make C# a more functional language they maybe focus on more functional feature (like discriminated unions) rather than needless changes like this.

**Reply by br3nt** [reply]
> Is the sole point of this just saving 2 brackets solely in case of single statement?

Sure, but it's more nuanced than that.  It's about the nicer development experience that comes from that.  That je ne sais quoi.  https://youtu.be/gls09kO-8DE.

This adds subtle flexibility, to existing language features that could help to improve how code is written.  Some other benefits are discussed below.

> Since it's introduction I've seen developers write convoluted code in order to needlessly make a method a single expression.
> If the plan is to make C# a more functional language they maybe focus on more functional feature (like discriminated unions) rather than needless changes like this.

This proposal isn't making C# more functional, rather it's enhancing and improving the current functionality.

"Since it's introduction I've seen developers write convoluted code in order to needlessly make a method a single expression" is a really good answer to why this isn't just a "needless change".

It adds the something that is missing from expression-bodied members.  It adds a nice syntax that brings many more language structures  to lambdas and method body expressions.

Sure, we can already use any of all these structures by introducing a block with `{` and `}`, but as you said, why would a dev prefer this over "reformat from an expression-bodied member to a regular method"?  So they make a choice to either refactor or make the statement more complex and convoluted to fit.

This proposal is a nuanced and subtle improvement, rather than in your face.  It's an obvious evolution that improves the dev experience and enhances the existing language features.  It will likely have a big impact on how devs write code because as you said, currently, "developers write convoluted code in order to needlessly make a method a single expression".  This proposal will go a long way to help alleviate that issue.

**Comment by HaloFour** [comment]
I have concerns that this syntax will limit any future conversations about whether or not said statements can also have expression forms, at least without requiring completely new syntax for those forms.

IMO, the shorthand isn't worth it.  You can already achieve the brevity by being less zealous about brace placement.

**Reply by CyrusNajmabadi** [reply]
> IMO, the shorthand isn't worth it. You can already achieve the brevity by being less zealous about brace placement.

Personally I agree.  I'll run this by LDM though to get their thoughts.  @KathleenDollard and @BillWagner likely have lots of thoughts here.

**Reply by br3nt** [reply]
Are expression forms very case-by-case? Why were expression forms not made for the other blocks at the same time as the switch expression?  Have other expressions already been considered/proposed?

**Reply by br3nt** [reply]
> You can already achieve the brevity by being less zealous about brace placement

Can you show an example for consideration?

**Comment by CyrusNajmabadi** [comment]
Moved to issue https://github.com/dotnet/csharplang/issues/6887.  My bad that i said to open another discussion.
